pkg/serverless/protocol.go

```diff
@@ -139,19 +147,58 @@ func (d *Daemon) TriggerFlush(ctx context.Context, shutdown bool) {
 	}
 }
 
+// Stop causes the Daemon to gracefully shut down. After a delay, the HTTP server
+// is shut down, data is flushed a final time, and then the agents are shut down.
+func (d *Daemon) Stop() {
+	// Can't shut down before starting
+	// If the DogStatsD daemon isn't ready, wait for it.
+	d.ReadyWg.Wait()
```

Comment:
I think this is the correct thing to do, however, I'm slightly concerned that a shutdown right after a failed startup may lock here if the daemon is never ready. 

I've checked and I think there is a scenario where it could happen if the trace agent fails to start here: https://github.com/DataDog/datadog-agent/blob/master/cmd/serverless/main.go#L359
This return should probably be removed to avoid having an issue in the future. 

I just want to emphase that using this lock in the shutdown can be tricky / slightly risky, but still, probably the right thing to do.

Reply:
I went ahead and removed the return.

I see your point but also note that the Lambda Platform will kill the extension after two seconds during shutdown.